File: JS/Tema9_Ejercicio7.js
/*
 * TEMA 9: PROGRAMACIÓN ORIENTADA A OBJETOS - EJERCICIO 7
 * ========================================================
 * Composición (asociación entre objetos).
 * 
 * EXPLICACIÓN:
 * - Composición: contener objetos dentro de otros
 * - Alternativa a herencia
 * - Relación "tiene un"
 */

class Motor {
    constructor(tipo, potencia) {
        this.tipo = tipo;
        this.potencia = potencia;
    }
    
    obtenerInfo() {
        return `${this.tipo} (${this.potencia} HP)`;
    }
}

class Llantas {
    constructor(marca, tamaño) {
        this.marca = marca;
        this.tamaño = tamaño;
    }
    
    obtenerInfo() {
        return `Llantas ${this.marca} ${this.tamaño}`;
    }
}

class Auto {
    constructor(marca, modelo) {
        this.marca = marca;
        this.modelo = modelo;
        this.motor = null;
        this.llantas = null;
    }
    
    instalarMotor(motor) {
        this.motor = motor;
    }
    
    instalarLlantas(llantas) {
        this.llantas = llantas;
    }
    
    obtenerEspecificaciones() {
        console.log(`\n${this.marca} ${this.modelo}`);
        console.log(`Motor: ${this.motor ? this.motor.obtenerInfo() : "No instalado"}`);
        console.log(`Llantas: ${this.llantas ? this.llantas.obtenerInfo() : "No instaladas"}`);
    }
}

class Oficina {
    constructor(nombre, area) {
        this.nombre = nombre;
        this.area = area;
        this.empleados = [];
    }
    
    agregarEmpleado(empleado) {
        this.empleados.push(empleado);
    }
    
    contarEmpleados() {
        return this.empleados.length;
    }
    
    mostrarInfo() {
        console.log(`\nOficina: ${this.nombre} (${this.area}m²)`);
        console.log(`Empleados (${this.contarEmpleados()}): ${this.empleados.join(", ")}`);
    }
}

console.log("=== COMPOSICIÓN ===");

const motorV8 = new Motor("V8", 350);
const llantas = new Llantas("Michelin", "18 pulgadas");

const carro = new Auto("Ferrari", "F8");
carro.instalarMotor(motorV8);
carro.instalarLlantas(llantas);
carro.obtenerEspecificaciones();

const oficina = new Oficina("Desarrolladores", 200);
oficina.agregarEmpleado("Juan");
oficina.agregarEmpleado("Ana");
oficina.agregarEmpleado("Carlos");
oficina.mostrarInfo();

/*
 * SALIDA ESPERADA:
 * === COMPOSICIÓN ===
 * 
 * Ferrari F8
 * Motor: V8 (350 HP)
 * Llantas: Llantas Michelin 18 pulgadas
 * 
 * Oficina: Desarrolladores (200m²)
 * Empleados (3): Juan, Ana, Carlos
 */

File: Python/Tema9_Ejercicio7.py
"""
TEMA 9: PROGRAMACIÓN ORIENTADA A OBJETOS
EJERCICIO 7: COMPOSICIÓN
==========================

Combina objetos (relación "tiene-un")
"""

# COMPOSICIÓN BÁSICA
print("=== COMPOSICIÓN BÁSICA ===\n")

class Motor:
    def __init__(self, tipo, potencia):
        self.tipo = tipo
        self.potencia = potencia
    
    def info(self):
        return f"{self.tipo} ({self.potencia} HP)"

class Auto:
    def __init__(self, marca, modelo):
        self.marca = marca
        self.modelo = modelo
        self.motor = None
    
    def instalar_motor(self, motor):
        self.motor = motor

auto = Auto("Toyota", "Corolla")
auto.instalar_motor(Motor("V6", 200))

print(f"{auto.marca} {auto.modelo}")
print(f"Motor: {auto.motor.info()}")

# COMPOSICIÓN COMPLEJA
print("\n=== COMPOSICIÓN COMPLEJA ===\n")

class Llantas:
    def __init__(self, marca, tamaño):
        self.marca = marca
        self.tamaño = tamaño

class AutoCompleto:
    def __init__(self, marca, modelo):
        self.marca = marca
        self.modelo = modelo
        self.motor = None
        self.llantas = []
    
    def instalar_motor(self, motor):
        self.motor = motor
    
    def instalar_llantas(self, cantidad, marca, tamaño):
        self.llantas = [Llantas(marca, tamaño) for _ in range(cantidad)]

auto = AutoCompleto("BMW", "X5")
auto.instalar_motor(Motor("V8TwinTurbo", 500))
auto.instalar_llantas(4, "Michelin", "20 pulgadas")

print(f"{auto.marca} {auto.modelo}")
print(f"Motor: {auto.motor.info()}")
print(f"Llantas: {len(auto.llantas)} x {auto.llantas[0].marca} {auto.llantas[0].tamaño}")

# OFICINA CON EMPLEADOS
print("\n=== OFICINA ===\n")

class Empleado:
    def __init__(self, nombre, cargo):
        self.nombre = nombre
        self.cargo = cargo

class Oficina:
    def __init__(self, nombre):
        self.nombre = nombre
        self.empleados = []
    
    def contratar(self, nombre, cargo):
        self.empleados.append(Empleado(nombre, cargo))
    
    def listar_empleados(self):
        for emp in self.empleados:
            print(f"  {emp.nombre} - {emp.cargo}")

oficina = Oficina("Tech Solutions")
oficina.contratar("Juan", "Desarrollador")
oficina.contratar("Ana", "Diseñadora")
oficina.contratar("Carlos", "Gerente")

print(f"Empleados en {oficina.nombre}:")
oficina.listar_empleados()

"""
SALIDA ESPERADA:
=== COMPOSICIÓN BÁSICA ===

Toyota Corolla
Motor: V6 (200 HP)

=== COMPOSICIÓN COMPLEJA ===

BMW X5
Motor: V8TwinTurbo (500 HP)
Llantas: 4 x Michelin 20 pulgadas

=== OFICINA ===

Empleados en Tech Solutions:
  Juan - Desarrollador
  Ana - Diseñadora
  Carlos - Gerente
"""
